README.md
What is this?
=============

Dimensions is a pattern-matching game.

How to use?
===========

Dimensions is not part of the Sugar desktop, but can be added.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [Download Dimensions using Browse](https://activities.sugarlabs.org/), search for `Dimensions`, then download, and;
* [How to use Dimensions](https://help.sugarlabs.org/dimensions.html)


NEWS
60
* Fix sub png images render bug
* L10 refresh, trim headers
* Update POT file
* License changes
* Fix card image persistence
* Port telepathy static bindings to TelepathyGLib
* Port to Python 3
* Update cairoplot
* Fix XML parse error
* Fix collaboration
* Avoid use of set_cursor

59

* New translations
* Added repository to activity.info

58

* High score chart (Ignacio Rodriguez)

57

* Reintegration of all three games: patterns, numbers, and words
* Simplification of toolbar
* New end of game animations
* Better feedback when playing computer (robot)
* Fixed regression in collaboration
* Fixed regression in scoring
* Fixed problem that cause cards to occasionally fly off the screen

56

* Reintegration of all three games: patterns, numbers, and words
* Simplification of toolbar
* New end of game animations
* Better feedback when playing computer (robot)
* Fixed regression in collaboration
* Fixed regression in scoring
* Fixed problem that cause cards to occasionally fly off the screen
* More simple mechanism for custom cards

55

* New translations

53

* Fix license text
* New translations

51

* Add support for screen rotation
* Make separate games for shapes, numbers, and words

49

* New translations

47

* New translations

45

* New translations

43

BUG FIX:
* Fixed name change in radio button (#4028)

41

ENHANCEMENTS:
* GTK-3 version
* Added summary to activity.info
* Added support for touch

39

BUG FIX:
* Closing the Welcome message clicks on a card too (#3567)
* Sometimes cards disappear (#3568)
* Detect and correct race condition (#3579)
(Thanks to Manuel Kaufmann for help tracking down these bugs)

38

ENHANCEMENTS:
* New translations
* Add stop button to help animation

BUG FIX:
http://dev.laptop.org/ticket/11804
* Refactoring of animation to use less memory
* Refactoring of deck to use less memory

37

BUG FIX:
* Fixed typo that prevented help animation from terminating

ENHANCEMENT:
* Autoadvance levels when play improves

36

ENHANCEMENTS:
* Support for dragging cards
* Visual feedback on mismatches
* Cursor when editing cards
* New help animation on first load
* Copy scores to clipboard (for import into Simple Graph)
* Notification when loading new games
* Slight delay after match before dealing new cards to allow for some
  scrutiny of results
* Addition of toolbar help menus
* Rebranding as Dimensions (a more apt name)
* Updated wiki page (See http://wiki.sugarlabs.org/go/Activities/Dimensions)
* New translations
(Tip of the hat to Mark Battley for great feedback and suggestions.)

35

ENHANCEMENTS:
* New translations

34

ENHANCEMENTS:
* New translations

BUG FIX:
* Fixed problem with animated cards missing destinations

33

ENHANCEMENTS:
* New translations
* New end-of-game graphics

32

ENHANCEMENT:
* Using Cairo graphics
* Animate loading of new cards

BUG FIX:
* Set robot time during "attract mode"

31

ENHANCEMENT:
* Animate matches when moving to discard pile

BUG FIX:
* FIx GPLv3 license address

30

ENHANCEMENTS:
* Switching buttons to radio buttons
* More button icon refactoring

29

ENHANCEMENT:
* "Standardized" icons

28

ENHANCEMENT:
* Added "attract mode" on first launch

27

ENHANCEMENTS:
* Smiley-face "reward" for finishing a game
* Improved workflow for importing custom cards

BUG FIXES:
* Work around datastore.find issue on 0.84 builds
* Work around chooser issues with old Sugar builds (#2398)

26

ENHANCEMENT:
* Import custom cards from Journal

25

ENHANCEMENTS:
* Quipu number game
* Optimization of sprite refresh

24

ENHANCEMENTS:
* Added beginner level
* Detect unsolvable game-play situations

23

BUG FIXES:
* Byron Corrales fixed the robot/word-edit conflict (#2057)
* General code cleanup

22

BUG FIX:
* Fixed SVG bug in old Sugar builds

21

BUG FIXES:
* general clean up of sprite library
* updated licenses

20

ENHANCEMENT:
* Added more combinations to the dice game, e.g., 5=4+1, 5=3+2

BUG FIXES:
* Fixed bug in sprite library
* Using refactored card generator

19

BUG FIXES:
* Fixed bug with Mayan 33
* Cleaned up some icons that had SVG problems on old Sugar builds

18

ENHANCEMENTS:
* Word-game lists can be edited
* Mayan number game
* "Fileless" SVG generation

BUG FIX:
* Fixed SVG problem on old Sugar builds

17

BUG FIXES:
* Fixed Class problem with command-line version
* Cleaned up license language (GPL v3)

16

ENHANCEMENTS:
* Simple sharing/collaborating
* Save/resume from Journal 
* New (improved?) toolbar layout
* Match-history animation at end game

BUG FIX:
* Fixed bug preventing play on old Sugar instances

15

ENHANCEMENTS:
* Added low/high levels for game play
* New activity icon

14

ENHANCEMENT:
* Computer assistant

13

ENHANCEMENT:
* More number games

12

ENHANCEMENTS:
* Reverting to three colors to simplify game play
* Experimenting with different number-game card styles
* Adding bold and italic to word cards
* Consolidate grid after dealing extra cards
* Refactoring of sprites library

11

BUG FIXES:
* Add check for match on first deal
* Fixed bug in extra card positions

10

ENHANCEMENTS:
* es, fr translation
* Rotated cards to enable larger/longer text strings

BUG FIX:
* Cleaned up match logic for word games

9

ENHANCEMENTS:
* Auto-deal of extra cards when needed
* Added glue for launching outside of Sugar
* Save "low score" to Journal
* Added numbers and words games
* Autogenerate cards on the fly
* Minutes:seconds display for total time
* General code cleanup

8

ENHANCEMENT:
* Added unselect

BUG FIX:
* Fixed bug in new game labels

7

ENHANCEMENTS:
* More distinct colors on cards
* Count seconds between matches
* Detect game over
* Automatically deals extra cards when necessary

6

BUG FIX:
* Fixed card selection bug

5

ENHANCEMENT:
* Added search function

4

ENHANCEMENT:
* Add 3 cards when you get stuck

3

ENHANCEMENTS:
* Eliminate naming alert from pre-0.86 toolbar
* New .pot file

2

ENHANCEMENT:
* New icon

BUG FIX:
* Fixed problem with the New Game button

1


constants.py
# -*- coding: utf-8 -*-
# Copyright (c) 2009, Walter Bender

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.

# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA


from gettext import gettext as _
LABELH = 32
MATCH_POSITION = 10
SHAPES = 3
COLORS = 3
NUMBER = 3
FILLS = 3
CARDS_IN_A_MATCH = 3
DECKSIZE = SHAPES * COLORS * NUMBER * FILLS
DEAL = 12
EXTRAS = 3
SELECTMASK = -1
MATCHMASK = -2
BACKGROUNDMASK = -3
CARD_WIDTH = 125
CARD_HEIGHT = 75
ROW = 5
COL = 3
KEYMAP = ['1', '2', '3', 'q', 'w', 'e', 'a', 's', 'd', 'z', 'x', 'c', '8',
          '9', '0']

CARD_STYLES = ['pattern', 'number', 'word', 'custom']
NUMBER_STYLES_O = ['roman', 'product', 'chinese', 'word', 'mayan', 'incan']
NUMBER_STYLES_C = ['lines', 'dots', 'star', 'hash', 'dice']

ROMAN = 0
PRODUCT = 1
CHINESE = 2
WORD = 3
MAYAN = 4
INCAN = 5
LINES = 0
DOTS = 1
STAR = 2
HASH = 3
DICE = 4

LOW = 'low'
MEDIUM = 'medium'
HIGH = 'high'
DIFFICULTY_LEVEL = [LOW, MEDIUM, HIGH]

RED_STROKE = '#FF6040'
RED_FILL = '#FFC4B8'
DARK_RED = '#802014'
BLUE_STROKE = '#0060C8'
BLUE_FILL = '#ACC8E4'
DARK_BLUE = '#003064'
GREEN_STROKE = '#00B418'
GREEN_FILL = '#AFE8A8'
DARK_GREEN = '#005A0C'
BLACK = '#000000'
WHITE = '#FFFFFF'
YELLOW = '#FFFF00'
GRAY = '#C0C0C0'
DARK_COLOR = [DARK_RED, DARK_GREEN, DARK_BLUE]
COLOR_PAIRS = ([RED_STROKE, RED_FILL],
               [GREEN_STROKE, GREEN_FILL],
               [BLUE_STROKE, BLUE_FILL])
FILL_STYLES = ['solid', 'none', 'gradient']
CARD_TYPES = ['X', 'O', 'C']
ROMAN_NUMERALS = {5: 'V', 7: 'VII', 10: 'X', 11: 'XI', 14: 'XIV', 15: 'XV',
                  21: 'XXI', 22: 'XXII', 33: 'XXXIII'}
NUMBER_NAMES = {5: _('five'), 7: _('seven'), 11: _('eleven'), 10: _('ten'),
                14: _('fourteen'), 15: _('fifteen'), 22: _('twenty two'),
                21: _('twenty one'), 33: _('thirty three')}
NUMBER_PRODUCTS = {5: '1×5', 7: '1×7', 11: '1×11', 10: '2×5',
                   14: '2×7', 15: '3×5', 22: '2×11',
                   21: '3×7', 33: '3×11'}
CHINESE_NUMERALS = {5: '五', 7: '七', 10: '十', 11: '十一', 14: '十四',
                    15: '十五', 21: '廿一', 22: '廿二', 33: '卅三'}
WORD_STYLES = ['font-weight: bold', '', 'font-style: italic']
WORD_CARD_INDICIES = [0, 4, 8, 36, 40, 44, 72, 76, 80, None, None, None, None,
                      None, None]
WORD_CARD_MAP = {WORD_CARD_INDICIES[0]: (0, 0), WORD_CARD_INDICIES[1]: (0, 1),
                 WORD_CARD_INDICIES[2]: (0, 2), WORD_CARD_INDICIES[3]: (1, 0),
                 WORD_CARD_INDICIES[4]: (1, 1), WORD_CARD_INDICIES[5]: (1, 2),
                 WORD_CARD_INDICIES[6]: (2, 0), WORD_CARD_INDICIES[7]: (2, 1),
                 WORD_CARD_INDICIES[8]: (2, 2)}
# CUSTOM_CARD_INDICIES = [0, 4, 8, 36, 40, 44, 72, 76, 80, None, None, None,
#                         None, None, None]
CUSTOM_CARD_INDICIES = [0, 1, 2, 36, 37, 38, 72, 73, 74, None, None, None,
                        None, None, None]

# 'dead key' Unicode dictionaries
DEAD_KEYS = ['grave', 'acute', 'circumflex', 'tilde', 'diaeresis', 'abovering']
DEAD_DICTS = [{'A': 192, 'E': 200, 'I': 204, 'O': 210, 'U': 217, 'a': 224,
               'e': 232, 'i': 236, 'o': 242, 'u': 249},
              {'A': 193, 'E': 201, 'I': 205, 'O': 211, 'U': 218, 'a': 225,
               'e': 233, 'i': 237, 'o': 243, 'u': 250},
              {'A': 194, 'E': 202, 'I': 206, 'O': 212, 'U': 219, 'a': 226,
               'e': 234, 'i': 238, 'o': 244, 'u': 251},
              {'A': 195, 'O': 211, 'N': 209, 'U': 360, 'a': 227, 'o': 245,
               'n': 241, 'u': 361},
              {'A': 196, 'E': 203, 'I': 207, 'O': 211, 'U': 218, 'a': 228,
               'e': 235, 'i': 239, 'o': 245, 'u': 252},
              {'A': 197, 'a': 229}]
NOISE_KEYS = ['Shift_L', 'Shift_R', 'Control_L', 'Caps_Lock', 'Pause',
              'Alt_L', 'Alt_R', 'KP_Enter', 'ISO_Level3_Shift', 'KP_Divide',
              'Escape', 'Return', 'KP_Page_Up', 'Up', 'Down', 'Menu',
              'KP_Home', 'KP_End', 'KP_Up', 'Super_L',
              'KP_Down', 'KP_Left', 'KP_Right', 'KP_Page_Down', 'Scroll_Lock',
              'Page_Down', 'Page_Up']
WHITE_SPACE = ['space', 'Tab']
